website/blog/2024/04-23-scrapy-vs-crawlee/index.md

```diff
@@ -0,0 +1,318 @@
+    }
+
+    ROTATING_PROXY_LIST = [
+        'proxy1.com:8000',
+        'proxy2.com:8031',
+        # Add more proxies as needed
+    ]
+```
+
+Now create a spider with the code you want to scrape any site and the `ROTATING_PROXY_LIST` in `settings.py` will manage which proxy to use for each request. Here middleware will treat each proxy initially as valid and then when a request is made, the middleware selects a proxy from the list of available proxies. The selection isn't purely sequential but is influenced by the recent history of proxy performance. The middleware has mechanisms to detect when a proxy might be banned or rendered ineffective. When such conditions are detected, the proxy is temporarily deactivated and put into a cooldown period. After the cooldown period expires, the proxy is reconsidered for use.
+
+In Crawlee, you can [use your own proxy servers](https://crawlee.dev/docs/guides/proxy-management) or proxy servers acquired from third-party providers. If you already have your proxy URLs, you can start using them like this:
+
+```js
+import { ProxyConfiguration } from 'crawlee';
+
+const proxyConfiguration = new ProxyConfiguration({
+    proxyUrls: [
+        'http://proxy1.example.com',
+        'http://proxy2.example.com',
+    ]
+});
+const crawler = new CheerioCrawler({
+    proxyConfiguration,
+    // ...
+});
+```
+
+Crawlee also has [`SessionPool`](https://crawlee.dev/api/core/class/SessionPool), a built-in allocation system for proxies. It handles the rotation, creation, and persistence of user-like sessions. It creates a pool of session instances that are randomly rotated.
+
+### Data Storage
+
+One of the most frequently required features when implementing scrapers is being able to store the scraped data as an "export file".
+
+Scrapy provides this functionality out of the box with the [`Feed Exports`](https://docs.scrapy.org/en/latest/topics/feed-exports.html), which allows it to generate feeds with the scraped items, using multiple serialization formats and storage backends. It supports `CSV, JSON, JSON Lines, and XML.`
+
+To do this, you need to modify your `settings.py` file and enter:
+
+```py
+    # To store in CSV format 
+    FEEDS = { 
+        'data/crawl_data.csv': {'format': 'csv', 'overwrite': True} 
+    }
+
+    # OR to store in JSON format
+
+    FEEDS = { 
+        'data/crawl_data.json': {'format': 'json', 'overwrite': True} 
+    }
+```
+
+Crawlee's storage can be divided into two categories: Request Storage (Request Queue and Request List) and Results Storage (Datasets and Key Value Stores). Both are stored locally by default in the `./storage` directory.
+
+Also, remember that Crawlee, by default, clears its storages before starting a crawler run. This action is taken to prevent old data from interfering with new crawling sessions.
+
+Let's see how Crawlee stores the result:
+
+- You can use local storage with dataset
+
+```js
```

Comment:
you might want to indent the whole codeblock to have it under the list item, now it will break the list (you will have a single item list, followed by codeblock, then another single item list and another codeblock)

and when i say indent the codeblock, i really mean indent it, not just its contents, the whole codeblock

https://meta.stackexchange.com/questions/3792/how-to-nest-code-within-a-list-using-markdown